src/drivers/esp8266_drv.c

```diff
@@ -309,211 +466,294 @@ static void init_wifi()
+                pr_error(LOG_PFX "Failed to register IPD callback\r\n");
 }
 
-static bool is_client_wifi_on_desired_network(const struct esp8266_client_info *ci)
+/**
+ * The method that checks our state machine and ensures that we are in
+ * the proper init state.  If we are not then the logic is here to get
+ * us to the correct state (if possible).
+ */
+static void check_init()
 {
-        if (!ci || !ci->has_ap || !state.client.config)
-                return false;
+        cmd_check_complete(CHECK_INIT);
+
+        pr_info(LOG_PFX "Checking Init\r\n");
+        switch(state.device.init_state) {
+        case DEV_INIT_STATE_READY:
+                /* Then we are where we want to be and are done */
+                return;
+        default:
+                /* STIEG: TODO Handle reset here when implemented. */
+                delayMs(CLIENT_BACKOFF_MS);
+        case DEV_INIT_STATE_NOT_READY:
+                init_wifi();
+                return;
+        }
+}
 
-        return STR_EQ(ci->ssid, state.client.config->ssid);
+static bool device_initialized()
+{
+        return state.device.init_state == DEV_INIT_STATE_READY;
 }
 
+
+/* *** Methods that handle the client WiFi state and actions *** */
+
+/**
+ * Callback that gets invoked with the get_client_ap command completes.
+ */
 static void get_client_ap_cb(bool status, const struct esp8266_client_info *ci)
 {
-        memcpy(&state.client.info, ci, sizeof(struct esp8266_client_info));
-        state.client.next_check_ms = date_time_uptime_now_plus(_CLIENT_BACKOFF_MS);
+        cmd_completed();
+        cmd_set_check(CHECK_CLIENT);
 
-        if (!status) {
-                /* Command failed */
-                cmd_completed(_CMD_UNKNOWN, 0);
-        } else if (!is_client_wifi_on_desired_network(ci)) {
-                /* Not on the network we want to be on.  Set it */
-                pr_info(_LOG_PFX "Client not on correct network\r\n");
-                cmd_completed(_CMD_CLIENT_AP_SET, 0);
-        } else {
-                /* On the network we want to be on.  Get the IP. */
-                pr_info(_LOG_PFX "Client is on desired network\r\n");
-                esp8266_log_client_info(&state.client.info);
-                cmd_completed(_CMD_CLIENT_IP_GET, 0);
-        }
+        memcpy(&state.client.info, ci, sizeof(struct esp8266_client_info));
+        *state.client.info.ip = 0;
 }
 
+/**
+ * Command that will populate our wifi client info with useful data.
+ */
 static void get_client_ap()
 {
+        pr_info(LOG_PFX "Retrieving Wifi Client Info\r\n");
         esp8266_get_client_ap(get_client_ap_cb);
         cmd_started();
 }
 
+/**
+ * Callback that gets invoked when the get_client_ip command completes.
+ */
 static void get_client_ip_cb(bool status, const char* ip)
 {
+        cmd_completed();
+        cmd_set_check(CHECK_CLIENT);
+
+        /* STIEG: This is a HACK.  IP Info should be its own struct */
```

Comment:
I do internally.  I have lots of stuff on an internal doc for what needs to get done.  I can file these in GH tickets if you think it will help.